plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -962,7 +1004,58 @@ void producer_plugin::plugin_initialize(const boost::program_options::variables_
       }
    }
 
-   my->_max_read_only_transaction_time_ms = options.at("max-read-only-transaction-time").as<int32_t>();
+   my->_ro_thread_pool_size = options.at( "read-only-threads" ).as<uint16_t>();
+   // only initialize other read-only options when read-only thread pool is enabled
+   if ( my->_ro_thread_pool_size > 0 ) {
+      EOS_ASSERT( my->_producers.empty(), plugin_config_exception, "--read-only-thread not allowed on producer node" );
+
+      if (chain.is_eos_vm_oc_enabled()) {
+         // EOS VM OC requires 4.2TB Virtual for each executing thread. Make sure the memory
+         // required by configured read-only threads does not exceed the total system virtual memory.
+         auto get_vm_sizes = []() -> std::pair<size_t, size_t> {
+            std::string attr_name;
+            size_t vm_total { 0 };
+            size_t vm_used { 0 };
+            std::ifstream meminfo_file("/proc/meminfo");
+            while (meminfo_file >> attr_name) {
+               if (attr_name == "VmallocTotal:") {
```

Comment:
I'm not sure this will behave as expected on a CPU that has five level paging enabled. `VmallocTotal` might report 64PB but `mmap()` will not give access to the extended range unless providing a hint (which currently we don't). Wish I had more time to test in an emulator to say for sure.

Reply:
It was agreed to set a hard limit of 8 threads for eos-vm-oc. Further work is tracked by https://github.com/AntelopeIO/leap/issues/801

Reply:
fwiw, I did boot a CPU will 5 level paging, and indeed `VmallocTotal:   13743895347199 kB`

However it seems pretty clear this isn't the value that we should be looking at anyways. This isn't reporting the virtual memory space available to user processes. On my desktop `VmallocTotal:   34359738367 kB` but that's just 32TB -- not 128TB. Instead, `VmallocTotal` appears to be the _kernel_ virtual address space available.

Reply:
Thanks Matt. Are you aware of any APIs which can be used to get the user space virtual memory available size? I did some searches and could not find anything useful.

Reply:
I'm not.. and a quick look I came up empty too. Maybe could probe around and try to discover where MAP_FIXED would allow you to go; but have to be mindful of existing mappings.

I think realistically we'll need to put a very high upper bound on the number of threads (1024?) and then refactor OC's usage to ensure in that worst case we're not pushing the envelope. Putting an upper bound on resource usage is important even for non-OC (set it to 100000 threads and see what happens :wink:). Obviously not a blocker for now though.

Reply:
Thanks. We will look into this further and put an upper limit on the number of threads in RC.